homeassistant/components/http/forwarded.py

```diff
@@ -0,0 +1,136 @@
+      - If malformed or invalid (IP) data in X-Forwarded-For header is found,
+        an HTTP 400 status code is thrown.
+      - The connected client peer on the socket of the incoming connection,
+        must be trusted for any processing to take place.
+    """
+    #
+    @middleware
+    async def forwarded_middleware(request, handler):
+        """Process forwarded data by a reverse proxy."""
+        overrides = {}
+
+        # Handle X-Forwarded-For
+        forwarded_for = request.headers.getall(X_FORWARDED_FOR, [])
+        if not forwarded_for:
+            # No forwarding headers, continue as normal
+            return await handler(request)
+
+        # Ensure the IP of the connected peer is trusted
+        connected_ip = ip_address(request.transport.get_extra_info("peername")[0])
+        if not any(connected_ip in trusted_proxy for trusted_proxy in trusted_proxies):
+            # Not trusted, continue as normal
+            return await handler(request)
+
+        # Multiple X-Forwarded-For headers
+        if len(forwarded_for) > 1:
+            _LOGGER.error("Too many headers for X-Forwarded-For", extra=request.headers)
+            raise HTTPBadRequest
+
+        # Process X-Forwarded-For from the right side (by reversing the list)
+        forwarded_for = list(reversed(forwarded_for[0].split(",")))
+        try:
+            forwarded_for = [ip_address(addr.strip()) for addr in forwarded_for]
+        except ValueError:
+            _LOGGER.error(
+                "Invalid IP address in X-Forwarded-For header", extra=request.headers
+            )
+            raise HTTPBadRequest
+
+        # Find the last trusted index in the X-Forwarded-For list
+        index = 0
+        for forwarded_ip in forwarded_for:
+            if any(forwarded_ip in trusted_proxy for trusted_proxy in trusted_proxies):
+                index += 1
+                continue
+            overrides["remote"] = str(forwarded_ip)
+            break
+
+        # If all the IP addresses are from trusted networks, take the left-most.
+        if "remote" not in overrides:
+            index = -1
+            overrides["remote"] = str(forwarded_for[-1])
+
+        # Handle X-Forwarded-Proto
+        forwarded_proto = list(reversed(request.headers.getall(X_FORWARDED_PROTO, [])))
+        if forwarded_proto:
+            if len(forwarded_proto) > 1:
+                _LOGGER.error(
+                    "Too many headers for X_FORWARDED_PROTO header",
+                    extra=request.headers,
+                )
+                raise HTTPBadRequest
+            forwarded_proto = forwarded_proto[0].split(",")
+            forwarded_proto = [p.strip() for p in forwarded_proto]
+
+            # Ideally this should take the scheme corresponding to the entry
+            # in X-Forwarded-For that was chosen, but some proxies only retain
+            # one element. In that case, use what we have.
```

Comment:
Shouldn't we raise a bad request unless this is either 1 (the last one) or the same length as the number of proxies used?

Reply:
That is a nice additional check that can be put in 👍

Reply:
Added in e518fc3, added tests in 0e3b587